### PR TITLE
perf(memory): batch re-embed backfill into one provider request

### PR DESCRIPTION
**Pull request by mysma-9403**
## Summary

- Collapse the re-embed backfill's per-item embedding into **one batched provider request** per bounded batch. `handle_reembed_backfill` previously embedded each chunk/summary body with a separate sequential `embedder.embed(&body).await` — up to `REEMBED_BACKFILL_BATCH = 16` real Voyage round-trips per batch, chained across the whole re-embed.
- Add a default `Embedder::embed_batch(&[&str]) -> Vec<Result<Vec<f32>>>` (one `Result` per input position) plus `embed_batch_via_provider`, which issues a single inner-provider `embed(&[..])` call and falls back to per-text `embed_one` on a whole-batch error or a length mismatch.
- Override `embed_batch` in `CloudEmbedder` and `OpenAiCompatEmbedder` (both wrap a batch-capable `EmbeddingProvider`); Ollama keeps the sequential default (its memory-tree path posts a single `prompt`, no batch endpoint).
- Refactor the two near-identical reembed loops into one generic `reembed_collect`. **No behavior change** beyond the embed call shape — all #1574 §6 failure/tombstone semantics and log lines are preserved verbatim.

## Problem

The genuine network-bound embedding work in memory-sync lives in the **re-embed backfill** path (extract-job / signature-switch driven), not the per-item ingest loop (which defers embedding to async extract jobs and only does CPU-bound fast scoring synchronously). `handle_reembed_backfill` reads each row's stored body and embeds it with a sequential `embedder.embed(&body).await`. For the shipped default (logged-in user, cloud `CloudEmbedder` → Voyage) every one of those is a real ~50–200ms network RTT, run strictly one after another up to 16 per batch, with the chain `Defer`-ing to revisit until the whole space is covered. After an embedder switch every prior row is missing at the new signature, so the backfill walks the entire memory tree this way.

## Solution

Split the embed call shape from the surrounding bookkeeping:

- **`Embedder::embed_batch` (new default trait method):** takes `&[&str]`, returns `Vec<Result<Vec<f32>>>` aligned by index (always `texts.len()` elements). The default is the old sequential loop, so any `Embedder` keeps working unchanged.
- **`embed_batch_via_provider`:** the shared override body. Issues **one** `inner.embed(&[..])` provider call (the inner `EmbeddingProvider` is natively batched), dim-checks each vector into its slot, and — on a whole-batch `Err` or a returned-length mismatch — **falls back to per-text `embed_one`** so a flaky batch endpoint degrades to the old behavior and per-position error attribution / tombstoning is never lost.
- **`CloudEmbedder` / `OpenAiCompatEmbedder`** override `embed_batch` to delegate to that helper. Ollama does not (no batch endpoint), so it keeps the sequential default.
- **`reembed_collect` (new generic helper):** folds the two copy-pasted loops (chunks, summaries) into one. Phase A reads bodies and persistently tombstones read failures; Phase B issues a single `embed_batch`; Phase C zips results back to ids and classifies each position **exactly** as the legacy loop did (`pack_checked` ok → keep, wrong dim → tombstone `"embed wrong dim"`, `Err(e)` → tombstone `"embed failed: {e}"`). A defensive length-mismatch guard skips the batch without tombstoning (so a later batch retries) rather than misattributing results.

The SQLite sidecar write (Phase 3, one short tx) is unchanged. Ordering is irrelevant — results fold into order-independent state keyed by id.

## Submission Checklist

- [x] Tests added or updated (happy path + at least one failure / edge case) — 7 new unit tests for the batch helper + default trait method; see Impact.
- [x] **Diff coverage ≥ 80%** — the new executable logic (`embed_batch`, `embed_batch_via_provider`, `reembed_collect`) is exercised directly by unit tests and the existing `handle_reembed_backfill` integration tests.
- [x] Coverage matrix updated — `N/A: internal perf change, no feature row affected`.
- [x] All affected feature IDs from the matrix are listed under `## Related` — `N/A`.
- [x] No new external network dependencies introduced — tests use an in-process `FakeProvider`/`SeqEmbedder`, no network.
- [x] Manual smoke checklist updated — `N/A: no release-cut surface touched (background memory-sync job only)`.
- [x] Linked issue closed via `Closes #NNN` — `N/A: no tracking issue`.

## Impact

- **Platform:** core (`openhuman` lib); background re-embed backfill job only. No UI/CLI surface change.
- **Performance:** the embed phase of a re-embed backfill batch goes from N serial Voyage round-trips to **one batched request** (N ≤ 16 per batch). Largest win on a post-switch full re-embed; steady state (small/no backfill) is unaffected. Under Ollama (no batch endpoint) or `embeddings_provider = "none"` the shape is unchanged — this is a background job, not interactive latency.
- **Parity:** `reembed_collect` preserves the #1574 §6 contract verbatim — worklist, tombstone reasons (`"body read failed: {e}"`, `"embed wrong dim"`, `"embed failed: {e}"`), log prefix `[memory::jobs] reembed_backfill:`, `Defer`/`Done` outcomes, and the per-signature "attempt-at-most-once" tombstone guarantee.
- **Tests:** `embed_batch_via_provider_happy_is_single_call` (one batch call, no per-text fallback), `_empty_makes_no_call`, `_falls_back_on_batch_error` (1 failed batch + N per-text), `_falls_back_on_length_mismatch`, `_maps_wrong_dim_per_position` (length matched → single call, bad slot maps to `Err`), plus `default_embed_batch_calls_embed_per_text` and `_preserves_per_position_errors`. Focused suites: embed **49 passed**, reembed **10 passed**.

## Related

- Closes: N/A
- Follow-up PR(s)/TODOs: N/A

---

### Linear Issue

- Key: N/A
- URL: N/A

### Commit & Branch

- Branch: `perf/reembed-backfill-batch`
- Commit SHA: `cff8188dd2aa6d0515d6486b9bb2e81c209267f6`

### Validation Run

- [x] `pnpm --filter openhuman-app format:check` — `N/A: no app/ (TS) changes`
- [x] `pnpm typecheck` — `N/A: no app/ (TS) changes`
- [x] Focused tests: `cargo test --lib memory_tree::score::embed` (49 passed), `cargo test --lib reembed` (10 passed)
- [x] Rust fmt/check (if changed): `cargo fmt` + `cargo check --lib` clean (exit 0)
- [x] Tauri fmt/check (if changed): `N/A: no app/src-tauri changes`

### Validation Blocked

- `command:` N/A
- `error:` N/A
- `impact:` N/A

### Behavior Changes

- Intended behavior change: embed-phase batching only (one provider request instead of N serial round-trips per backfill batch).
- User-visible effect: faster re-embed backfill (notably after an embedder switch) when using a batch-capable cloud embedder; no change to what gets embedded or stored.

### Parity Contract

- Legacy behavior preserved: worklist queries, `Defer`/`Done` outcomes, `REEMBED_BACKFILL_BATCH` bound, per-row read/embed log-and-skip, persistent tombstones (`mark_chunk_reembed_skipped` / `mark_summary_reembed_skipped`) with identical reason strings, sidecar write tx, progress/summary logs.
- Guard/fallback/dispatch parity checks: `embed_batch` whole-batch-error and length-mismatch both fall back to per-text `embed_one`; `reembed_collect` length-mismatch guard skips without tombstoning so rows stay re-embeddable; Ollama keeps the sequential default (no batch endpoint).

### Duplicate / Superseded PR Handling

- Duplicate PR(s): N/A
- Canonical PR: N/A
- Resolution (closed/superseded/updated): N/A

## Summary by CodeRabbit

* **New Features**
  * Batched embedding for bulk text processing, used during re-embedding/backfill to improve throughput and efficiency.
  * Per-item handling that records skipped items when embedding fails or dimensions mismatch.

* **Bug Fixes**
  * Enforced strict embedding-dimension validation to avoid silent misalignments and surface per-item errors.

* **Tests**
  * Added deterministic async tests covering batch happy-paths, fallbacks, and per-item error cases.

### CODE REVIEW

File: src/openhuman/memory_queue/handlers/mod.rs [hunk 1]
```diff
@@ -703,6 +703,89 @@ fn try_mark_summary_reembed_skipped(
...
+        // violation means we can't attribute results to ids — skip the whole
+        // batch WITHOUT tombstoning so a later batch retries these rows.
+        log::error!(
+            "[memory::jobs] reembed_backfill: {label} embed_batch returned {} results for {} texts; \
+             skipping batch without tombstoning (sig={active_sig})",
+            results.len(),
+            readable.len()
+        );
+        return Vec::new();
+    }
```
[the remark this answers is not in the document]
Comment (mysma-9403) [reply src/openhuman/memory_queue/handlers/mod.rs:765]: Fixed in e8f25b54. `reembed_collect` now returns `Result<…>` and `anyhow::bail!`s on the length-contract violation instead of returning an empty `Vec`. Because `handle_reembed_backfill` already returns `Result<JobOutcome>` and both call sites are now `.await?`, the error propagates and terminates the chain rather than spinning on an endless `Defer` that re-selects the same ids forever. Also added `embed_batch:enter/:success/:fallback` debug telemetry to the batch path so the single-round-trip win vs. the per-text fallback is traceable.
